arangod/Cluster/ClusterCollectionMethods.cpp

```diff
@@ -693,6 +707,39 @@ LOG_TOPIC("e16ec", WARN, Logger::CLUSTER)
   return shardNames;
 }
 
+[[nodiscard]] auto ClusterCollectionMethods::prepareCollectionGroups(
+    ClusterInfo& ci, std::string_view databaseName,
+    std::vector<CreateCollectionBody> const& collections)
+    -> ResultT<replication2::CollectionGroupUpdates> {
+  arangodb::replication2::CollectionGroupUpdates groups;
+  std::unordered_map<std::string, replication2::agency::CollectionGroupId>
+      selfCreatedGroups;
+  for (auto const& col : collections) {
+    if (col.distributeShardsLike.has_value()) {
+      auto const& leadingName = col.distributeShardsLike.value();
+      if (selfCreatedGroups.contains(leadingName)) {
+        groups.addToNewGroup(selfCreatedGroups.at(leadingName), col.id);
+      } else {
+        // TODO: This code needs to look-up the CollectionID.
+        // It is not yet added as part of the Collection Properties.
+        auto c = ci.getCollection(databaseName, leadingName);
+        TRI_ASSERT(c.get() != nullptr);
+        // We never get a nullptr here because an exception is thrown if the
+        // collection does not exist. Also, the createCollection should have
+        // failed before.
+        groups.addToExistingGroup(
+            replication2::agency::CollectionGroupId{c->id().id()}, col.id);
+      }
+    } else {
+      // Create a new CollectionGroup
+      auto groupId = groups.addNewGroup(col);
+      // Remember it for reuse
+      selfCreatedGroups.emplace(col.name, groupId);
+    }
+  }
```

Comment:
Is it possible that the collection referenced through `leadingName` is located after the current `col` in the `collections` vector? It won't have any `distributeShardsLike` and will be added to `seldCreatedGroups`, but that would be too late, since we would need it already during the processing of current `col`. Just checking.